spec/unit/lib/delayed_job/threaded_worker_spec.rb

```diff
@@ -0,0 +1,175 @@
+    end
+  end
+
+  describe '#start' do
+    before do
+      allow(worker).to receive(:threaded_start)
+      allow(worker.instance_variable_get(:@mutex)).to receive(:synchronize).and_call_original
+    end
+
+    it 'sets up signal traps for all signals' do
+      expect(worker).to receive(:trap).with('TERM')
+      expect(worker).to receive(:trap).with('INT')
+      expect(worker).to receive(:trap).with('QUIT')
+      worker.start
+    end
+
+    it 'starts the specified number of threads' do
+      expect(worker).to receive(:threaded_start).exactly(num_threads).times
+
+      worker.start
+
+      expect(worker.instance_variable_get(:@threads).length).to eq(2)
+      expect(worker.instance_variable_get(:@mutex)).to have_received(:synchronize).twice
+    end
+
+    it 'logs the start and shutdown messages' do
+      expect(worker).to receive(:say).with('Starting threaded delayed worker with 2 threads')
+      worker.start
+    end
+
+    it 'sets the thread_name variable for each thread' do
+      worker.start
+      worker.instance_variable_get(:@threads).each_with_index do |thread, index|
+        expect(thread[:thread_name]).to eq("thread:#{index + 1}")
+      end
+    end
+
+    it 'logs the error and stops the worker when an unexpected error occurs' do
+      allow(worker).to receive(:threaded_start).and_raise(StandardError.new('test error'))
+      allow(worker).to receive(:stop)
+      expect { worker.start }.to raise_error('Unexpected error occurred in one of the worker threads')
+      expect(worker.instance_variable_get(:@unexpected_error)).to be true
+    end
+  end
+
+  describe '#name' do
+    it 'returns the instance name if thread name is set' do
+      allow(Thread.current).to receive(:[]).with(:thread_name).and_return('some-thread-name')
+      expect(worker.name).to eq('instance_name some-thread-name')
+    end
+  end
+
+  describe '#stop' do
+    it 'logs the shutdown message' do
+      queue = Queue.new
+      allow(worker).to(receive(:say)) { |message| queue.push(message) }
+
+      worker.stop
+      expect(queue.pop).to eq('Shutting down worker threads gracefully...')
+    end
+
+    it 'sets the exit flag in the parent worker' do
+      worker.stop
+      sleep 0.1 until worker.instance_variable_get(:@exit)
+      expect(worker.instance_variable_get(:@exit)).to be true
+    end
+
+    it 'allows threads to finish their work without being killed prematurely' do
+      allow(worker).to receive(:threaded_start) do
+        5.times do
+          break if worker.instance_variable_get(:@exit)
+
+          sleep 0.5
+        end
+      end
+
+      worker_thread = Thread.new { worker.start }
+      sleep(0.5)
+      expect(worker.instance_variable_get(:@threads).all?(&:alive?)).to be true
+      worker.instance_variable_get(:@threads).each { |t| allow(t).to receive(:kill).and_call_original }
+
+      Thread.new { worker.stop }.join
+      worker.instance_variable_get(:@threads).each(&:join)
```

Comment:
This `join` is done internally in the ThreadedWorker and should not be repeated in the test.